### PR TITLE
Add support for ASN.1 tags on multiple bytes

### PR DESCRIPTION
**Pull request by frankmorgner**
I am using the ASN.1 parser to read card verifiable certificates used in ICAO, for example. They use tags on more than one byte, like `0x7F4E` for the certificate body. This PR adds support for parsing and writing tags with more than one byte.

I am actively using the parser code. The writer code should work, but is untested so far. All internal tests are passing (`make test`).

### CODE REVIEW

**Comment by frankmorgner** [comment]
successfully tested `mbedtls_asn1_write_tag`.

File: include/mbedtls/asn1write.h [hunk 1]
```diff
@@ -54,8 +54,7 @@ int mbedtls_asn1_write_len( unsigned char **p, unsigned char *start, size_t len
  *
  * \return          the length written or a negative error code
  */
-int mbedtls_asn1_write_tag( unsigned char **p, unsigned char *start,
-                    unsigned char tag );
+int mbedtls_asn1_write_tag( unsigned char **p, unsigned char *start, int tag );
```
Comment (sbutcher-arm) [inline include/mbedtls/asn1write.h]: This is a breaking change to the interface unfortunately. To introduce the functionality I think you want, you'll need to create a new function.

File: library/asn1write.c [hunk 2]
```diff
@@ -99,12 +99,15 @@ int mbedtls_asn1_write_len( unsigned char **p, unsigned char *start, size_t len
     return( MBEDTLS_ERR_ASN1_INVALID_LENGTH );
 }
 
-int mbedtls_asn1_write_tag( unsigned char **p, unsigned char *start, unsigned char tag )
+int mbedtls_asn1_write_tag( unsigned char **p, unsigned char *start, int tag )
```
Comment (sbutcher-arm) [inline library/asn1write.c]: Again, this is a breaking change to the interface, so we can't accept the PR as it is. You'll need to create a new function.

**Comment by sbutcher-arm** [comment]
Unfortunately the PR needs some work before we can accept it. I haven't done a proper review, but I can see in it's current form we can't accept it. If you'd like to look again, and resubmit the PR, we'd welcome the contribution.

Also, we have a policy of only accepting contributions with a  Contributor’s Licence Agreement (CLA), signed or authorised by yourself, so to progress the contribution we really need a CLA.

The easiest way to do this is if you create an mbed account and [accept this click through agreement](https://developer.mbed.org/contributor_agreement/) which is applicable if this is a personal contribution. Alternatively, you can find a [slightly different agreement to sign here](https://www.mbed.com/en/about-mbed/contributor-license-agreements/), which can be signed and returned to us, and is applicable if you don't want to create an mbed account or alternatively if this is a corporate contribution.

Thanks for your understanding!

**Comment by frankmorgner** [comment]
I refactored the change to keep the old interfaces and accepted the CLA as https://os.mbed.com/users/frankmorgner/.

File: library/asn1write.c [hunk 3]
```diff
@@ -101,10 +101,18 @@ int mbedtls_asn1_write_len( unsigned char **p, unsigned char *start, size_t len
...
+
+int mbedtls_asn1_write_long_tag( unsigned char **p, unsigned char *start, int tag )
+{
+    while ( tag != 0 ) {
+        if( *p - start < 1 )
+            return( MBEDTLS_ERR_ASN1_BUF_TOO_SMALL );
 
-    *--(*p) = tag;
+        *--(*p) = tag & 0xFF;
+        tag >>= 8;
```
[the remark this answers is not in the document]
Comment (frankmorgner) [reply library/asn1write.c]: Well, the original API doesn't check whether the tag that fits `unsigned char` is a valid tag that can be written on one byte. Following this scheme, I didn't add any validation or modification of the tag supplied. I agree that validating the tag is useful, but that would have nothing to do with what I'm adressing here and should be done in a seperate pull request.

Regarding your suggestion to use tags that are magically encoded with the correct flags is a bit more tricky:
1. Many specifications use the raw tags, which may lead to problems when bits get added magically
2. mbedtls doesn't provide any means to translate a raw tag into a human readable one. For example, in OpenSSL you could write a short raw tag `0x53` as `ASN1_APP_IMP(CVC_CHAT_SEQ, relative_authorization, ASN1_OCTET_STRING, 0x13)`. In mbedtls, however, there aren't any short cuts for the implicit application specific tag `0x13` so that the caller `mbedtls_asn1_write_tag` already has to translate this to the raw value `0x53`.

In summary, my implementation sticks to the current paradigms by choosing simplicity over too much magic.

File: library/asn1write.c [hunk 4]
```diff
@@ -101,10 +101,18 @@ int mbedtls_asn1_write_len( unsigned char **p, unsigned char *start, size_t len
...
+    }
 
     return( 1 );
```
[the remark this answers is not in the document]
Comment (frankmorgner) [reply library/asn1write.c]: OK, thanks for the pointer

File: include/mbedtls/asn1write.h [hunk 5]
```diff
@@ -44,6 +44,14 @@ extern "C" {
  */
 int mbedtls_asn1_write_len( unsigned char **p, unsigned char *start, size_t len );
 
+/**
+ * \brief           Write a ASN.1 tag in ASN.1 format
+ *
+ * \note            See \c mbedtls_asn1_write_long_tag for description
```
[the remark this answers is not in the document]
Comment (frankmorgner) [reply include/mbedtls/asn1write.h]: Everything is identical to `mbedtls_asn1_write_long_tag`. Do you want me to copy that description?

**Comment by frankmorgner** [comment]
Thanks for the review, @andresag01. I've updated the PR with the changes requested and explained the paradigms [above](https://github.com/ARMmbed/mbedtls/pull/848#discussion_r199153024).

- Unfortunately, I currently don't have time to dive deeper into the test cases. But if you add some boiler plate for testing the standard cases, I'd be happy to add some more test data with long tags.
- tags on two bytes are the longest I've seen (e.g. in ICAO's card verifiable certificates). I think `int` is well enough.

**Comment by andresag01** [comment]
@frankmorgner: Thank you for reworking the PR and apologies for the delay in my reply. I have added the `needs design approval` tag as this patch effectively adds a new feature.

**Comment by frankmorgner** [comment]
Squashed into a single commit and created https://github.com/ARMmbed/mbed-crypto/pull/386 to meet the recent separation within mbedtls.

**Comment by gilles-peskine-arm** [comment]
Thank you for your contribution! Unfortunately, we can't accept this new feature without tests. Please add unit tests for multibyte tags.

**Comment by daverodgman** [comment]
Closing as this needs additional work as discussed above - but please feel free to re-open if you have time to update this PR.

**Comment by frankmorgner** [comment]
Just to get this right: I've reworked the code to fix all of the above comments regarding the actual implementation and documentation, but you don't want to include this because of some missing test cases?

That's unfortunate and I wish I had known this requirement from the start to save me the work I put into this :disappointed:

**Comment by gilles-peskine-arm** [comment]
Mbed TLS, like many other software projects, especially in high-reliability fields such as cryptography, generally doesn't accept new code without tests. We do have historical test gaps in some places, and you may have been misled because back when you submitted this pull request ASN.1 was not unit-tested directly (it is now), only via X.509 parsing. If so please accept our apologies. Nonetheless, the new feature _does_ need tests. Ideally a few unit test cases for the new parser case and the new write function, and at least one positive test case with a certificate from your application domain.

**Comment by frankmorgner** [comment]
That's fine and understood, but it was not clear when I've opened this PR *5 years ago* nor when it was first reviewd *one year later*.

I could provide example data for parsing CV certificates with longer tags as used for MRTD, but I fear that writing some unittest especially for the encoder would need some work. So for now, I'm leaving this as is.
